### PR TITLE
Add external URL linking card and modal

### PR DESCRIPTION
**Pull request by ianrowan**
## Link to Issue
Closes: #4829 

## Description of Changes
- Adds support for `LinkSource.Web` type links via a link card on view thread page
- Includes a modal for creating and deleting new URL links
- Includes ability to title links 

image

image
image

## Test Plan
- Go to a thread where you are an author or admin
- Click on 'manage Links' in the new card
- Add a link of your choice and select +, add as many links as desired
- Ensure links populate and work on thread page

- Click manage links 
- De-select a link you wish to delete and save changes
- ensure link is deleted

## Other Considerations

- See conversation in ticket. There is an additional(potentially) user story to tag 'proposal external links' which is still in discussion and can be planned to be included in this ticket(probably a 2 point addition)
- One work around idea was that users can title links anything with these changes so the user story could be achieved in the meantime without tags

### CODE REVIEW

File: packages/commonwealth/client/scripts/views/pages/view_thread/linked_url_card.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import { Link } from 'react-router-dom';
+import app from 'state';
+import { CWSpinner } from 'views/components/component_kit/cw_spinner';
+import type Thread from '../../../models/Thread';
+import { LinkSource } from '../../../models/Thread';
+import { CWButton } from '../../components/component_kit/cw_button';
+import { CWContentPageCard } from '../../components/component_kit/CWContentPage';
+import { Modal } from '../../components/component_kit/cw_modal';
+import { CWText } from '../../components/component_kit/cw_text';
+import { LinkedUrlModal } from '../../modals/linked_url_modal';
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/pages/view_thread/linked_url_card.tsx:13]: linked_url_modal => LinkedUrlModal

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx [hunk 2]
```diff
@@ -52,6 +52,7 @@ import { LinkedThreadsCard } from './linked_threads_card';
 import { LockMessage } from './lock_message';
 import { ThreadPollCard, ThreadPollEditorCard } from './poll_cards';
 import { SnapshotCreationCard } from './snapshot_creation_card';
+import { LinkedUrlCard } from './linked_url_card';
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:55]: linked_url_card => LinkedUrlCard

**Review by masvelio — COMMENTED** [review]
left some small comments, please fix file namings. Otherwise lgtm. 
Did not do click around testing

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx (hunk 2 above)
Comment (ianrowan) [reply packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx:55]: used this as it looked like the standard for all of the other cards
image

File: packages/commonwealth/client/scripts/views/pages/view_thread/linked_url_card.tsx (hunk 1 above)
Comment (ianrowan) [reply packages/commonwealth/client/scripts/views/pages/view_thread/linked_url_card.tsx:13]: See other comment

File: packages/commonwealth/client/scripts/views/components/url_link_selector/url_selector.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  const handleInputChange = (e) => {
+    setSearchTerm(e.target.value);
+  };
+
+  const EmptyComponent = () => (
+    <div className="empty-component">{getEmptyContentMessage()}</div>
+  );
+
+  return (
+    <div className="SnapshotProposalSelector">
```
[the remark this answers is not in the document]
Comment (ianrowan) [reply packages/commonwealth/client/scripts/views/components/url_link_selector/url_selector.tsx]: updated this

File: packages/commonwealth/client/scripts/views/components/url_link_selector/url_selector.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import { UrlSelectorItem } from './url_selector_item';
+
+type UrlSelectorProps = {
+  onSelect: (link: Link) => void;
+  urlsToSet: Pick<Link, 'identifier'>[];
+};
+
+export const UrlSelector = ({ onSelect, urlsToSet }: UrlSelectorProps) => {
+  const [searchTerm, setSearchTerm] = useState('');
+  const [loading, setLoading] = useState(false);
```
[the remark this answers is not in the document]
Comment (ianrowan) [reply packages/commonwealth/client/scripts/views/components/url_link_selector/url_selector.tsx]: removed

File: packages/commonwealth/client/scripts/views/components/url_link_selector/index.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+import { UrlSelector } from './url_selector';
+import { UrlSelectorItem } from './url_selector_item';
```
[the remark this answers is not in the document]
Comment (ianrowan) [reply packages/commonwealth/client/scripts/views/components/url_link_selector/index.ts:2]: see other comment

File: packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx (hunk 2 above)
Comment (ianrowan) [reply packages/commonwealth/client/scripts/views/pages/view_thread/ViewThreadPage.tsx]: fixed per doc

File: packages/commonwealth/client/scripts/views/components/url_link_selector/url_selector.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (ianrowan) [reply packages/commonwealth/client/scripts/views/components/url_link_selector/url_selector.tsx]: fixed per doc

File: packages/commonwealth/client/scripts/views/components/url_link_selector/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ianrowan) [reply packages/commonwealth/client/scripts/views/components/url_link_selector/index.ts]: fixed per doc

**Review by masvelio — APPROVED** [review]
LG 🙌

**Comment by jnaviask** [comment]
- Search for URLs doesn't seem to do anything.
- UX wise, we may want to open in a new tab?
- would love to get a design pass on this to ensure that the UI is ready for production.

**Comment by ianrowan** [comment]
@jnaviask updated the links to open in a new tab. 

As for the search, I carried that over from other modals but thinking about it UX-wise I think we can remove it. Reasoning being that its typically used to search for unlinked proposals, etc whereas we're only using that querylist to show the created external links which I would guess wont be extensive enough typically to need to search to remove(the main reason you'd be looking around that list

**Comment by jnaviask** [comment]
@ianrowan feel free to remove the search here -- it exists for proposals e.g. because you're searching from the list of extant proposals which to add as links. In this case we have no need for it.

**Comment by ianrowan** [comment]
@jnaviask removed search, will await design on the rest

**Comment by jnaviask** [comment]
@ianrowan can we merge to demo + post in the product-design channel asking for reviews?

**Comment by ianrowan** [comment]
@jnaviask Yeah sure, whats the demo merge process? A separate PR?

**Comment by jnaviask** [comment]
@ianrowan the `demo` branch should theoretically auto-deploy and anyone can push, but definitely check with @Miaplacidus before pushing to it, to ensure you're not steamrolling someone else's work.

**Comment by jessmart1213** [comment]
### Design review

So I'm not quite sure how to review this modal as there are no designs to reference and Im not fond of the UX (please see [loom](https://www.loom.com/share/4cb35fdf5a4c4e9babc93ee6a98f6eb0?sid=83b9cae7-733f-4000-b5c6-e6184159c65f) for more details). That said, I understand that this modal probably following the UX pattern set by previous modals - so I won't officially write feedback here - will wait until its time to revamp this page.

**UI Feedback**
- Update modal to have 16px spacing between inputs. Currently these inputs have 0px spacing and it looks broken.

- This may need to be a separate rip & replace ticket. Update the modal UI to use the [new Design System modal](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=388%3A4508&mode=design&t=rD59QO2cyBUUSYMN-1)
- Apply to cards Linked Proposals, Linked Discussions, Web links, and Add poll.

**Comment by ianrowan** [comment]
Thanks for the review @jessmart1213. I think the modal can be a placeholder as this issue is aimed to unblock other work related to being able to have linked URLs on threads. 

I updated the spacing as you mentioned and agree that as part of another issue this and other link modals should be redesigned/updated. 

@jnaviask that said I think we could move forward with this for now to unblock the issues dependent on the ability to add webpage links to thread data + view in general

**Comment by zakhap** [comment]
@jnaviask Who is going to be responsible for updating the design requests that @jessmart1213 requested?

**Comment by masvelio** [comment]
@zakhap if this will be merged, I can take care of the design feedback. I will be working ([taking over Daniel's work](https://github.com/hicommonwealth/commonwealth/pull/4757))on modals in upcoming days, so I will make sure that every modal looks good

**Comment by jnaviask** [comment]
@masvelio would you be willing to clean up this PR in place per design feedback? A little hesitant to ship without explicit product approval.

**Comment by masvelio** [comment]
@jnaviask I am not sure what I supposed to do here? 

Per @jessmart1213 feedback 

> UI Feedback
> 1. Update modal to have 16px spacing between inputs. Currently these inputs have 0px spacing and it looks broken.
> 2. This may need to be a separate rip & replace ticket. Update the modal UI to use the [new Design System modal](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=388%3A4508&mode=design&t=rD59QO2cyBUUSYMN-1)
> 3. Apply to cards Linked Proposals, Linked Discussions, Web links, and Add poll.

Number 1 has been applied by @ianrowan.
Number 2 & 3 cannot be done within this PR. Not sure what component @zakhap referred to [here](https://github.com/hicommonwealth/commonwealth/pull/4876#issuecomment-1745247973), but new modal is not merged yet so we do not have it in component kit and we cannot simply use it. This can be done only after #4757 will land on master.

tldr:
1. either we merge it as it is and in upcoming days I will apply new modal layout for all modals in the app, including one from this PR
2. or we draft it until new modals will be merged to master and then this PR would need to be updated to use new modals

**Comment by jnaviask** [comment]
Thanks for the feedback @masvelio. Let's get this merged now to unblock @timolegros and we'll update its UX later on.
